libraries/networking/src/webrtc/WebRTCDataChannels.cpp

```diff
@@ -0,0 +1,492 @@
+    jsonCandidate.insert("sdpMLineIndex", candidate->sdp_mline_index());
+
+    QJsonObject jsonWebRTCData;
+    jsonWebRTCData.insert("candidate", jsonCandidate);
+
+    QJsonObject jsonObject;
+    jsonObject.insert("from", QString(_parent->getNodeType()));
+    jsonObject.insert("to", _webSocketID);
+    jsonObject.insert("data", jsonWebRTCData);
+    QJsonDocument jsonDocument = QJsonDocument(jsonObject);
+
+#ifdef WEBRTC_DEBUG
+    qCDebug(networking_webrtc) << "7. Send ICE candidate to the remote peer";
+#endif
+    _parent->sendSignalingMessage(jsonObject);
+}
+
+void WDCConnection::onDataChannelOpened(rtc::scoped_refptr<DataChannelInterface> dataChannel) {
+#ifdef WEBRTC_DEBUG
+    qCDebug(networking_webrtc) << "WDCConnection::onDataChannelOpened() :"
+        << dataChannel->id()
+        << QString::fromStdString(dataChannel->label())
+        << QString::fromStdString(dataChannel->protocol())
+        << dataChannel->negotiated()
+        << dataChannel->maxRetransmitTime()
+        << dataChannel->maxRetransmits()
+        << dataChannel->maxPacketLifeTime().value_or(-1)
+        << dataChannel->maxRetransmitsOpt().value_or(-1);
+#endif
+
+    _dataChannel = dataChannel;
+    _dataChannelID = _parent->getNewDataChannelID();  // Not dataChannel->id() because it's only unique per peer connection.
+    _dataChannel->RegisterObserver(_dataChannelObserver.get());
+
+    _parent->onDataChannelOpened(this, _dataChannelID);
+}
+
+void WDCConnection::onDataChannelStateChanged() {
+    auto state = _dataChannel->state();
+#ifdef WEBRTC_DEBUG
+    qCDebug(networking_webrtc) << "WDCConnection::dataChannelStateChanged() :" << (int)state
+        << DataChannelInterface::DataStateString(state);
+#endif
+    if (state == DataChannelInterface::kClosed) {
+        _dataChannel->Close();
+        _dataChannel = nullptr;
+        // WEBRTC FIXME: The following line causes the _peerConnectionFactory to fail.
+        //_peerConnection->Close();
+        //_peerConnection = nullptr;
+        _parent->onDataChannelClosed(this, _dataChannelID);
+    }
+}
+
+void WDCConnection::onDataChannelMessageReceived(const DataBuffer& buffer) {
+#ifdef WEBRTC_DEBUG
+    qCDebug(networking_webrtc) << "WDCConnection::onDataChannelMessageReceived()";
+#endif
+
+    auto byteArray = QByteArray(buffer.data.data<char>(), (int)buffer.data.size());
+
+    // Echo message back to sender.
+    if (byteArray.startsWith("echo:")) {
+#ifdef WEBRTC_DEBUG
+    qCDebug(networking_webrtc) << "Echo message back";
+#endif
+        _parent->sendDataMessage(_dataChannelID, byteArray);  // Use parent method to exercise the code stack.
+        return;
+    }
+
+    _parent->emitDataMessage(_dataChannelID, byteArray);
+}
+
+bool WDCConnection::sendDataMessage(const DataBuffer& buffer) {
+#ifdef WEBRTC_DEBUG
+    qCDebug(networking_webrtc) << "WDCConnection::sendDataMessage()";
+#endif
+    const int MAX_WEBRTC_BUFFER_SIZE = 16 * 1024 * 1024;  // 16MB
```

Comment:
Constant probably should be somewhere else, perhaps at the top of the file?
Why 16M? Is this a protocol issue, or just a sane memory limit?

What kind of data will be sent? Textures may exceed 16M quite easily.

There should be some sort of log message here, too.

Reply:
This is a WebRTC protocol limit. This is just how much data can be buffered to send. Ultimately, it's called via `Socket::writeDatagram()` - in PR1259 - and UDP sockets can similarly fail to write date. The Vircadia protocol handles this situation (dropping unreliable packets and resending reliable packets). Large data items are chopped up into multi-packet messages.